openmdao/core/problem.py

```diff
@@ -759,90 +760,167 @@ def _compute_total_derivs(self, of=None, wrt=None, return_format='flat_dict',
+
+        for rhs_name, vois in iteritems(voi_lists):
+            voi_info = {}
+            old_size = None
+
             # If Forward mode, solve linear system for each 'wrt'
             # If Adjoint mode, solve linear system for each 'of'
-            for icount, input_name in enumerate(input_list):
-                # Dictionary access returns a scaler for 1d input, and we
-                # need a vector for clean code, so use _views_flat.
-                flat_view = dinputs._views_flat[input_name]
+            for input_name, old_input_name in vois:
+                vecname = v2rhs_name[input_name]
+                dinputs = input_vec[vecname]
+                doutputs = output_vec[vecname]
+
                 in_var_idx = model._var_allprocs_abs2idx['output'][input_name]
                 in_var_meta = model._var_allprocs_abs2meta['output'][input_name]
-                start = np.sum(model._var_sizes['output'][:iproc, in_var_idx])
-                end = np.sum(model._var_sizes['output'][:iproc + 1, in_var_idx])
+                start = np.sum(sizes[:iproc, in_var_idx])
+                end = np.sum(sizes[:iproc + 1, in_var_idx])
 
                 in_idxs = None
                 if input_name in input_vois:
                     in_voi_meta = input_vois[input_name]
                     if 'indices' in in_voi_meta:
                         in_idxs = in_voi_meta['indices']
 
-                distrib = in_var_meta['distributed']
+                dup = not in_var_meta['distributed']
                 if in_idxs is not None:
                     irange = in_idxs
-                    loc_size = len(in_idxs)
-                    dup = False
+                    loc_size = new_size = len(in_idxs)
                 else:
-                    irange = range(in_var_meta['global_size'])
+                    new_size = in_var_meta['global_size']
+                    irange = list(range(new_size))
                     loc_size = end - start
-                    dup = not distrib
 
-                loc_idx = -1
-                for idx in irange:
+                if old_size is None:
+                    old_size = new_size
+                elif old_size != new_size:
+                    raise RuntimeError("Indices within the same VOI group must be "
+                                       "the same size, but in the group %s, %d != %d" %
+                                       (vecname, old_size, new_size))
+
+                if loc_size == 0:
+                    # var is not local. get size of var in owned proc
+                    for rank in range(nproc):
+                        sz = sizes[rank, in_var_idx]
+                        if sz > 0:
+                            loc_size = sz
+                            break
+
+                voi_info[input_name] = (dinputs, doutputs, irange, loc_size, start, end, dup)
+
+            loc_idxs = defaultdict(lambda: -1)
+
+            # at this point, we know that for all vars in the current
```

Comment:
The logic behind that is that we're potentially making collective MPI calls inside of model.solve_linear, so in order to keep everything in sync we have to iterate the same number of times in all of the procs.